src/operator/subgraph/dnnl/dnnl_remove_casts_property.h

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file dnnl_remove_casts_property.h
+ * \brief Graph property for removing two unnecessary Cast operations
+ *
+ * ... -> Cast(bool) -> expand_dims -> Cast(bool) -> Cast(bool) -> ...
+ *                                  ||
+ *                                  \/
+ *                ... -> Cast(bool) -> expand_dims -> ...
+ */
+
+#ifndef MXNET_OPERATOR_SUBGRAPH_DNNL_DNNL_REMOVE_CASTS_PROPERTY_H_
+#define MXNET_OPERATOR_SUBGRAPH_DNNL_DNNL_REMOVE_CASTS_PROPERTY_H_
+
+#if MXNET_USE_ONEDNN == 1
+
+#include <map>
+#include <string>
+#include <vector>
+
+#include "operator/subgraph/common.h"
+#include "dnnl_subgraph_base-inl.h"
+
+namespace mxnet {
+namespace op {
+
+class SgDNNLRemoveCastsSelector : public SubgraphSelectorV2 {
+ private:
+  enum CastStatus { kStart, kExpand, kCast, kSuccess, kFail };
+  CastStatus status_  = kStart;
+  const char* boolStr = "bool";
+
+ public:
+  bool Select(const BiDirectedNode& seed_node,
+              const std::shared_ptr<NodeAttr>& node_attr) override {
+    if (seed_node.node->op() == Op::Get("expand_dims") && seed_node.node->num_inputs() == 1 &&
+        seed_node.node->num_outputs() == 1) {
+      status_ = kExpand;
+      return true;
+    }
+    return false;
+  }
+
+  bool SelectInput(const BiDirectedNode& n, const BiDirectedNode& input_node) override {
+    if (input_node.node->op() != Op::Get("Cast")) {
+      status_ = kFail;
+    } else if (input_node.node->attrs.dict["dtype"] != boolStr) {
```

Comment:
Could it will be done of dtype's? Probably we can save dtype in Select and use it for input and output selector

Reply:
Select selects expand_dims which does not have dtype. This fuse is very specific therefore I did not even try to make it more generic.

Reply:
So maybe you can  notice the type of the cast in SelectInput and check it in SelectOutput (or vice versa)

Reply:
I will, thanks for your input :)